### ISSUE TITLE
IP stack: Incorrect algorithm for TCP sequence number comparison 

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1961]
**_Reported by Paul Sokolovsky:_**

tcp.c has a certain function:

{code:java}
/* True if the (signed!) difference "seq1 - seq2" is positive and less
 * than 2^29.  That is, seq1 is "after" seq2.
 */
static inline bool seq_greater(u32_t seq1, u32_t seq2)
{
        int d = (int)(seq1 - seq2);
        return d > 0 && d < 0x20000000;
}
```

However, nothing in https://tools.ietf.org/html/rfc793 talks about "29 bits" and stuff like that. The sequence number is clearly defined as:

```
Sequence Number:  32 bits
```

No references to 29 bits, no 0x20000000's.

Actually, there's a venerable document from 1978: https://www.rfc-editor.org/ien/ien74.txt. If human language may lie, then PDP-10 assembly and BCPL expressions given there can't go wrong.

If that's not enough, lwIP and Linux do it with plain 32bit arithmetics, no 29 bits either.

The function above is traceable to:

```
commit ddda50c214f7e0ac06a272873a912094883afcb7
Author: Jukka Rissanen <jukka.rissanen@linux.intel.com>
Date:   Thu Aug 18 10:08:04 2016 +0300

    net: tcp: Initial TCP functionality
    
    This creates initial TCP handling logic but does not yet
    enable fully working TCP connection.
    
    Some of the connection logic is taken from FNET TCP
    implementation.
    
    Origin: FNET 3.6.1
    URL: https://github.com/butok/FNET/blob/master/fnet_stack/stack/fnet_tcp.c
    
    Change-Id: I1e100d9fa9c91437562b933d94d0bd3db1a5885e
    Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>
```

Indeed, it comes from FNET, albeit the header file: https://github.com/butok/FNET/blob/master/fnet_stack/stack/fnet_tcp.h#L168 . As it stands, it appears to be "original research" in Wikipedia terms on the part of FNET authors.

Using incorrect sequence comparison algorithm certainly violates TCP protocol. It may also pose security issues (an obvious one might be possibility to quickly fingerprint FNET devices in the wild to exploit further vulnerabilities which it likely has a lot if even sequence comparison is done wrong).

Let's fix it.


(Imported from Jira ZEP-2120)

**Comment by nashif** [comment]
**_by Jukka Rissanen:_**

This functionality came from FNET as you found out. I agree with you, we should fix this.

**Comment by nashif** [comment]
**_by Jukka Rissanen:_**

I looked the seq_greater() function and isn't it working just fine actually. It checks that seq2 is greater than seq1 and also check if we wrapped (the d < 0x2000000) part.
What is actually missing is that the function would need to be available to net_context.c and we should use it there to check the sequence numbers.
Actually I already introduced tcp_validate_seq() function like that in PR#166 https://github.com/zephyrproject-rtos/zephyr/pull/166 Hmm, that func is probably not what we want here, sorry for the noise.
